java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/PromoService.java

```diff
@@ -0,0 +1,142 @@
+package net.osmand.server.api.services;
+
+import net.osmand.server.api.repo.DeviceSubscriptionsRepository;
+import net.osmand.server.api.repo.PremiumUsersRepository;
+import net.osmand.server.api.repo.PromoCampaignRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+
+import javax.transaction.Transactional;
+import java.util.*;
+
+@Service
+public class PromoService {
+    
+    @Autowired
+    PromoCampaignRepository promoCampaignRepository;
+    
+    @Autowired
+    private EmailSenderService emailSender;
+    
+    @Autowired
+    private PremiumUsersRepository usersRepository;
+    
+    @Autowired
+    protected UserSubscriptionService userSubService;
+    
+    @Autowired
+    private DeviceSubscriptionsRepository subscriptionsRepository;
+    
+    public void register(PromoCampaignRepository.Promo promo) {
+        if (promoCampaignRepository.existsById(promo.name)) {
+            throw new IllegalStateException("Promo already exists");
+        } else {
+            promoCampaignRepository.save(promo);
+        }
+    }
+    
+    @Transactional
+    public ResponseEntity<String> addUser(String name, String email) {
+        PromoCampaignRepository.Promo promoCampaign = promoCampaignRepository.findByName(name);
+        if (promoCampaign.used < promoCampaign.numberLimit) {
+            String key = "promo_" + promoCampaign.name;
+            Date expireTime = getExpirationDate(promoCampaign);
+            PromoResponse resp = createPromoSubscription(email, key, expireTime);
+            PremiumUsersRepository.PremiumUser existingUser = usersRepository.findByEmail(email);
+            if (existingUser != null && !resp.error) {
+                promoCampaign.used++;
+                promoCampaign.lastUsers = getLastUsers(promoCampaign);
+                promoCampaignRepository.save(promoCampaign);
+                return ResponseEntity.ok(expireTime.toString());
+            } else {
+                return ResponseEntity.badRequest().body(resp.deviceSub.purchaseToken);
+            }
+        }
+        return ResponseEntity.badRequest().body("Unfortunately we ran out of available promocodes");
+    }
+    
+    private String getLastUsers(PromoCampaignRepository.Promo promoCampaign) {
+        StringBuilder res = new StringBuilder();
+        List<DeviceSubscriptionsRepository.SupporterDeviceSubscription> subscriptions = subscriptionsRepository.findFirst5BySkuOrderByStarttimeDesc("promo_" + promoCampaign.name);
+        for (DeviceSubscriptionsRepository.SupporterDeviceSubscription subscription : subscriptions) {
+            PremiumUsersRepository.PremiumUser user = usersRepository.findByOrderid(subscription.orderId);
```

Comment:
add if (subscription.starttime > now() - 1 month) so we don't pull data for years back